[onnx] create a new port

luncliff opened the pull request:
### What does your PR fix?  

Create a new port for https://github.com/onnx/onnx. 

#### Related Issues?

* https://github.com/microsoft/vcpkg/pull/14903#issuecomment-825774151
* #17199 (`libtorch` PR)

#### To Do

* ONNXIFI with Facebook Extension https://github.com/houseroad/foxi --> will be next work after this port

### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  

Targets normal desktop environments with vcpkg's CI check

* `x64-windows`
* `x86-windows`
* `x64-osx`
* `x64-linux`

### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  

Yes

luncliff committed:
[onnx] create a new port

luncliff committed:
[onnx] update baseline

luncliff committed:
[onnx] fix pybind11 build in Linux

luncliff committed:
[onnx] fix windows build

* rewrite cmake patch
* provide CRT related options

luncliff committed:
[onnx] use Python_EXECUTABLE

luncliff committed:
[onnx] run codegen targets explicitly

* Disable UWP support

luncliff committed:
[onnx] fix python script install

luncliff committed:
[onnx] remove Development component in Windows

talregev commented:
@luncliff Need any help?

luncliff committed:
[onnx] finalize port works

talregev commented:
Did you fix everything? 
You can run tests by type:
/azp run

luncliff commented:
> Did you fix everything?
> You can run tests by type:
> /azp run
> 
> Thank you for the commits!

Thanks for check. I couldn't work on libtorch port dependencies, but starting it again.
Import worked fine. Several build fix commit will be added

talregev commented:
> > Did you fix everything?
> > You can run tests by type:
> > /azp run
> > Thank you for the commits!
> 
> Thanks for check. I couldn't work on libtorch port dependencies, but starting it again.
> Import worked fine. Several build fix commit will be added

Thank you!

luncliff committed:
[onnx] resurrect 47a8e8e

* fix version SHA

luncliff commented:
Found the error is from `tensorflow-cc`, Bazel version issue.
https://dev.azure.com/vcpkg/public/_build/results?buildId=56904&view=artifacts

luncliff commented:
/azp run

talregev commented:
@JackBoosY Can you help?

NancyLi1013 commented:
@luncliff 
Is this ready for review now?

luncliff commented:
I think so.

On Mon, Aug 2, 2021 at 5:48 PM NancyLi1013 ***@***.***> wrote:

> @luncliff <https://github.com/luncliff>
> Is this ready for review now?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/vcpkg/pull/18073#issuecomment-890847082>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADLSYE7AMH4WZS5QICI4JQTT2ZLVBANCNFSM45LUOKGA>
> .
>

NancyLi1013 commented on ports/onnx/portfile.cmake:
```diff
@@ -0,0 +1,100 @@
+vcpkg_find_acquire_program(PYTHON3)
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
```
```suggestion
    SOURCE_PATH "${SOURCE_PATH}"
```
The path should be quoted in cmake file. Please see the doc https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/cmake-guidelines.md.

luncliff replied:
Got it. I will check the other works either.

NancyLi1013 commented on ports/onnx/portfile.cmake:
```diff
@@ -0,0 +1,100 @@
+vcpkg_copy_pdbs()
+vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/ONNX)
+
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```
```suggestion
file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```

NancyLi1013 commented on ports/onnx/portfile.cmake:
```diff
@@ -0,0 +1,100 @@
+                    ${CURRENT_PACKAGES_DIR}/include/onnx/defs/controlflow
+                    ${CURRENT_PACKAGES_DIR}/include/onnx/defs/rnn
+                    ${CURRENT_PACKAGES_DIR}/include/onnx/defs/nn
+)
```
Could you please also add double quotes for these paths?

luncliff replied:
Yeah. I will double check the folders and quote them soon

NancyLi1013 commented on ports/onnx/vcpkg.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "onnx",
+  "version": "1.9.0",
```
```suggestion
  "version-semver": "1.9.0",
```
It might be better served by [version-semver](https://github.com/microsoft/vcpkg/blob/master/docs/users/versioning.md#version-semver).

luncliff committed:
Update ports/onnx/portfile.cmake

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

luncliff committed:
Update ports/onnx/portfile.cmake

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

luncliff committed:
Update ports/onnx/vcpkg.json

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

luncliff committed:
[onnx] rewrite portfile, patch

* add host dependencies for protoc/python3
* rewrite patch files to work with FindPython3

luncliff commented:
Added `host` dependencies for `protoc`, `python3` which is required for codegen steps.

My Linux environment kept using `/usr/bin/python3` which has lower version.
To fix this, I replaced `find_package(Python)` to `find_package(Python3)` and provide `Python3_ROOT_DIR=${CURRENT_INSTALLED_DIR}` to work correctly with `python3` port of vcpkg.

Hope this approach is correct.

luncliff committed:
[onnx] exclude onnxifi targets

* use more correct detection of protobuf shared libs

luncliff committed:
[onnx] use HOST_INSTALLED_DIR for python3

luncliff committed:
[onnx] handle Development component of Python3

* split the use of find_package(Python3)

update patch files

* More comment in portfile.cmake

NancyLi1013 commented:
Have you tested the feature `pybind11`? @luncliff

NancyLi1013 commented:
LGTM now, thanks for adding this new port. @luncliff

BillyONeal commented on ports/onnx/portfile.cmake:
```diff
@@ -0,0 +1,98 @@
+        pybind11 BUILD_ONNX_PYTHON
+)
+
+# Like protoc, python is required for codegen. Hijack the Python3_EXECUTABLE
+vcpkg_find_acquire_program(PYTHON3)
```
```suggestion
# Like protoc, python is required for codegen.
vcpkg_find_acquire_program(PYTHON3)
```

Is this required even when pybind11 is off?

luncliff replied:
Yes. protobuf and python is required for internal protobuf related code generation steps.

BillyONeal commented on ports/onnx/vcpkg.json:
```diff
@@ -0,0 +1,34 @@
+      "name": "protobuf",
+      "host": true
+    },
+    {
+      "name": "python3",
+      "host": true
+    },
```
Is this being added because of `vcpkg_find_acquire_program(PYTHON3)`, or is there some other use?

luncliff replied:
The part is for `find_package(Python3 COMPONENTS Interpreter)`. And for cross compile scenario, it uses `find_package(Python3 COMPONENTS Development)`. The dependency is solved by `pybind11`.

luncliff committed:
Update ports/onnx/portfile.cmake

Co-authored-by: Billy O'Neal <bion@microsoft.com>

luncliff committed:
[onnx] update version SHA
